File: Test/EscortDateRoles.m
#import "Kiwi.h"
#import "NSDate+Escort.h"
#import "NSDate+AZDateBuilder.h"
#import "FakeDateUtil.h"

@interface NSDate (EscortMock)
+ (NSCalendar *)AZ_currentCalendar;
@end

SPEC_BEGIN(EscortDateRoles)
    registerMatchers(@"AZ");// NSDate Custom Matcher
    describe(@"-isTypicallyWorkday", ^{
        __block NSDate *currentDate;
        beforeEach(^{
            currentDate = [NSDate AZ_dateByUnit:@{
                AZ_DateUnit.year : @2010,
                AZ_DateUnit.month : @10,
                AZ_DateUnit.day : @10,
            }];
            [FakeDateUtil stubCurrentDate:currentDate];
        });
        context(@"when Workday is first", ^{
            __block NSDate *firstDayOfWeek;
            beforeEach(^{
                firstDayOfWeek = [currentDate AZ_dateByUnit:@{
                    AZ_DateUnit.weekday : @1,
                }];
            });
            it(@"should be false", ^{
                BOOL match = [firstDayOfWeek isTypicallyWorkday];
                [[theValue(match) should] beNo];
            });
        });
        context(@"when Workday is last", ^{
            __block NSDate *lastDayOfWeek;
            beforeEach(^{
                lastDayOfWeek = [currentDate AZ_dateByUnit:@{
                    AZ_DateUnit.weekday : @7,
                }];
            });
            it(@"should be false", ^{
                BOOL match = [lastDayOfWeek isTypicallyWorkday];
                [[theValue(match) should] beNo];
            });
        });
        context(@"when Workday is typically workday", ^{
            __block NSRange weekdayRange;
            beforeEach(^{
                NSCalendar *calendar = [NSCalendar currentCalendar];
                weekdayRange = [calendar maximumRangeOfUnit:NSCalendarUnitWeekday];
            });
            it(@"should be true", ^{
                NSDate *checkDate;
                NSUInteger length = weekdayRange.location + weekdayRange.length - 1;
                for (NSUInteger weekday = weekdayRange.location + 1; weekday < length; weekday++) {
                    checkDate = [[currentDate AZ_dateByUnit:@{
                        AZ_DateUnit.weekday : @1,
                    }] dateByAddingDays:weekday - 1];
                    BOOL match = [checkDate isTypicallyWorkday];
                    [[theValue(match) should] beTrue];
                }
            });
        });
    });
    describe(@"-isTypicallyWeekend", ^{
        __block NSDate *currentDate;
        beforeEach(^{
            currentDate = [NSDate AZ_dateByUnit:@{
                AZ_DateUnit.year : @2010,
                AZ_DateUnit.month : @10,
                AZ_DateUnit.day : @10,
            }];
            [FakeDateUtil stubCurrentDate:currentDate];
        });
        context(@"when Weekday is first", ^{
            __block NSDate *firstDayOfWeek;
            beforeEach(^{
                firstDayOfWeek = [currentDate AZ_dateByUnit:@{
                    AZ_DateUnit.weekday : @1
                }];
            });
            it(@"should be true", ^{
                BOOL match = [firstDayOfWeek isTypicallyWeekend];
                [[theValue(match) should] beYes];
            });
        });
        context(@"when Weekday is last", ^{
            __block NSDate *lastDayOfWeek;
            beforeEach(^{
                lastDayOfWeek = [currentDate AZ_dateByUnit:@{
                    AZ_DateUnit.weekday : @7
                }];
            });
            it(@"should be true", ^{
                BOOL match = [lastDayOfWeek isTypicallyWeekend];
                [[theValue(match) should] beYes];
            });
        });
        context(@"when Weekday is typically workday", ^{
            __block NSRange weekdayRange;
            beforeEach(^{
                NSCalendar *calendar = [NSCalendar currentCalendar];
                weekdayRange = [calendar maximumRangeOfUnit:NSCalendarUnitWeekday];
            });
            it(@"should be true", ^{
                NSDate *checkDate;
                NSUInteger length = weekdayRange.location + weekdayRange.length - 1;
                for (NSUInteger weekday = weekdayRange.location + 1; weekday < length; weekday++) {
                    checkDate = [[currentDate AZ_dateByUnit:@{
                            AZ_DateUnit.weekday : @1,
                    }] dateByAddingDays:weekday - 1];
                    BOOL match = [checkDate isTypicallyWeekend];
                    [[theValue(match) should] beNo];
                }
            });
        });
    });
    SPEC_END

File: SwiftTest/EscortDateRoles.swift
import XCTest
import Quick
import AZDateBuilder

class EscortDateRolesSpec: QuickSpec {
    override func spec() {
        describe("isTypicallyWorkday") {
            it("when weekday is first should return false") {
                let currentDate = Date.date(by: [
                    .year: 2010,
                    .month: 10,
                    .day: 10,
                    ])
                XCTAssertFalse(currentDate.isTypicallyWorkday())
            }
            it("when weekday is last should return false") {
                let currentDate = Date.date(by: [
                    .year: 2010,
                    .month: 10,
                    .day: 16,
                    ])
                XCTAssertFalse(currentDate.isTypicallyWorkday())
            }
            it("when weekday is last should return true") {
                let currentDate = Date.date(by: [
                    .year: 2010,
                    .month: 10,
                    .day: 11 + Int(arc4random_uniform(5)),
                    ])
                XCTAssertTrue(currentDate.isTypicallyWorkday())
            }
        }
        describe("isTypicallyWeekend") {
            it("when weekday is first should return true") {
                let currentDate = Date.date(by: [
                    .year: 2010,
                    .month: 10,
                    .day: 10,
                    ])
                XCTAssertTrue(currentDate.isTypicallyWeekend())
            }
            it("when weekday is last should return true") {
                let currentDate = Date.date(by: [
                    .year: 2010,
                    .month: 10,
                    .day: 16,
                    ])
                XCTAssertTrue(currentDate.isTypicallyWeekend())
            }
            it("when weekday is last should return false") {
                let currentDate = Date.date(by: [
                    .year: 2010,
                    .month: 10,
                    .day: 11 + Int(arc4random_uniform(5)),
                    ])
                XCTAssertFalse(currentDate.isTypicallyWeekend())
            }
        }
    }
}
